src/bin/tikv-ctl.rs

```diff
@@ -182,3 +303,127 @@ fn dump_range(db: DB, from: String, to: Option<String>, limit: Option<u64>, cf:
                  })
         .unwrap()
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::str;
+    use rocksdb::Writable;
+    use tikv::util::codec::bytes::encode_bytes;
+    use tikv::util::codec::number::NumberEncoder;
+    use tikv::raftstore::store::keys;
+    use tikv::storage::{ALL_CFS, CF_LOCK, CF_WRITE, CF_DEFAULT};
+    use tikv::storage::mvcc::{Lock, Write, LockType, WriteType};
+    use tempdir::TempDir;
+    use tikv::util::rocksdb::new_engine;
+
+    const PREFIX: &'static [u8] = b"k";
+
+    #[test]
+    fn test_ctl_mvcc() {
+        let tmp_dir = TempDir::new("mvcc_tmp").expect("create mvcc_tmp dir");
+        let file_path = tmp_dir.path().join("tmp_db");
+        let db = new_engine(file_path.to_str().unwrap(), ALL_CFS).unwrap();
+        let test_data = vec![(PREFIX, b"v", 5), (PREFIX, b"x", 10), (PREFIX, b"y", 15)];
+        let keys: Vec<_> = test_data.iter()
+            .map(|data| {
+                let encoded = encode_bytes(&data.0[..]);
+                let mut d = keys::data_key(encoded.as_slice());
+                let _ = d.encode_u64_desc(data.2);
+                d
+            })
+            .collect();
+        let default_value: Vec<_> = test_data.iter().map(|data| *data.1).collect();
+        let kvs = keys.iter().zip(default_value.iter());
+        for (k, v) in kvs.clone() {
+            db.put(k.as_slice(), v).unwrap();
+        }
+        let kvs_gen: Vec<MvccKv<Vec<u8>>> = gen_mvcc_iter(&db, "k", CF_DEFAULT).unwrap();
+        let mut test_iter = test_data.clone();
+        assert_eq!(test_iter.len(), test_data.len());
+        for kv in kvs_gen {
+            let ts = kv.key.decode_ts().unwrap();
+            let key = kv.key.truncate_ts().unwrap().raw().unwrap();
+            let value = kv.value.as_slice();
+            test_iter.retain(|&s| !(&s.0[..] == key.as_slice() && &s.1[..] == value && s.2 == ts));
+        }
+        assert_eq!(test_iter.len(), 0);
+
+        // Test MVCC Lock
+        let test_data_lock = vec![(PREFIX, LockType::Put, b"v", 5),
+                                  (PREFIX, LockType::Lock, b"x", 10),
+                                  (PREFIX, LockType::Delete, b"y", 15)];
```

Comment:
Each key should has one lock at most. Here the 3 locks share the same key, the last lock will override others.
You need to setup 3 keys if you want to test all types of locks.